### ISSUE TITLE
Design Matrix diagnostic module

### ISSUE THREAD

**Issue by adelavega** [poldracklab/fitlins#213]
It would be useful to have a module of diagnostic checks on design matrices, like other analysis software provides.

For example, things such as checking for rank-deficiency of the design-matrix.

This is something I planned to add to neuroscout, but it makes more sense to put it somewhere generalization.

Since pybids will now only output a `df` and not a `design_matrix` per se, @tyarkoni suggested putting this module here.

The idea is to have a `Diagnostic` class, with something like a `run` function that takes a design matrix. The class will also have many functions starting with `check_` and the `run` function will apply all of them. We could have on/off flags in the `run` function to disbale some checks, or simply just return a long list of warnings that the user can feel free to ignore.

These can then be rendered on Neuroscout's frontend by importing from fitlins, as well as rendered by fitlin reports.

**Comment by effigies** [comment]
I think you may have more of this API in your head than I do. All I have is:

```Python
class Diagnostic:
    def run(self, design_matrix : pd.DataFrame):
        self.check_A(design_matrix)
        ...
        self.check_Z(design_matrix)

    def check_A(self, design_matrix):
        ...

    ...

    def check_Z(self, design_matrix):
        ...
```

And I guess the checks probably need to return something, and `run()` needs to return a list of those things, for the sake of reporting.

Do you have more details in mind? Is FitLins directly generating reports, or data structures that NeuroScout will turn into reports?

**Comment by adelavega** [comment]
Yeah, something like that. I think I'm going to start simple and simply issue a list of strings with warnings, if any. But we could get more sophisticated down the line.

**Comment by adelavega** [comment]
The latter, just returning data structures. Our reports are generated on a web app, so we wouldn't want fitlins to have to worry about, or cater to that. Your "reports" can use their own logic with these data structures as well. Ideally, they somewhat parallel each other in functionality, but not necessarily in aesthetics.

**Comment by adelavega** [comment]
Ideas for diagnostics:
- Warn about variables correlated greater than some threshold (default 0.95)
- Rank deficiency
- Design efficiency
- Warn about `n/a`s in final design matrix

**Comment by effigies** [comment]
> * Warn about variables correlated greater than some threshold (default 0.95)

Do we care if two nuisance regressors correlate highly? Assuming we know the variables of interest, we presumably only want to warn if a variable of interest correlates highly with some other regressor. I guess if we have contrasts for design efficiency, then we can "know" these. Or just return pairs and it's the job of the caller to make distinctions.

> * Rank deficiency
> * Design efficiency (requires contrasts)
> * Warn about `n/a`s in final design matrix

:+1:

> Any others?

Can't think of any off the top of my head.

**Comment by adelavega** [comment]
I think unless a DM is rank deficient, near rank deficiency is OK, in terms of estimability. So for nuisance variables, it doesn't matter much if they are highly correlated with each other. 

Highly correlated variables will require a lot of power though, so it would be a problem for variables of interest, and I think it's useful to know if a variable of interest is highly correlated with a nuisance regressor.

I don't like the idea of having to track which variables are of interest or not though. Instead, we could leave it up to the user to inspect the design efficiency / correlation matrices, and only warn if the DM is rank deficient can will probably crash. Or like you said, warn for all pairs above 0.95 regardless.

**Comment by adelavega** [comment]
Now I'm thinking the Diagnostic module should just make all plots / run all checks, and return a dictionary with everything. You could have flags that disable / configure some of these calls, and maybe make the methods @staticmethods so they could be used individually as well. 

The generation of specific messages can then be left up to the report generator (i.e. fitlins or neuroscout).
